### DISCUSSION TITLE
M03 doesn't wait until rpm reached

### DISCUSSION THREAD

**Discussion by ToastyCH** [grblHAL/core#198] [Ideas]
Hello togheter,

Im useing a diy milling machine.
Board grblhal2k, vfd hy-620 connected via rs485

When I start a gcode like this:

M03 S24000
G0 X0 Y0
G01 Z-5 F500

My machine starts the spindle but continues imediatley. So my insert reaches the workpiece before reaching 24k rpm. (I can see the spindle ramping up in ioSender)

In my eyes the right behavior would be that grblhal waits till the spindle reaches the desired rpm. And the continue with the gcode.

I created an github Account to post this idea. So I have no idea how everything works here. If i can provide more data, please let me know.

Best regards

Toasty

**Comment by LoganFraser** [comment]
Wait for AtSpeed rpm detection would be ideal, but even a post-processor addition of the spindle delay setting after *every* rpm increase would be useful, especially in ops that have different ramping to milling RPM settings.

**Comment by terjeio** [comment]
Setting `$340` is available? If so set it to the deviation allowed as a percentage of the programmed speed. `$392` can be used to set the delay before an alarm is raised if the spindle fails to reach the speed.

**Reply by Dietz0r** [reply]
> Setting `$340` is available? If so set it to the deviation allowed as a percentage of the programmed speed. `$392` can be used to set the delay before an alarm is raised if the spindle fails to reach the speed.

Spindle on delay is available and set appropiately but does not delay gcode execution. 
It just checks for RPM at the set time. I can reproduce this on VFD with RS485 and on PWM mode.

It was surmised that it's a conflict of the "Saftey Door Input" function.

**Reply by terjeio** [reply]
If `$340` is set to 0 the `$392` delay is ignored (for M3/M4 commanded spin up it is the delay before an alarm is raised, for safety door it is delay before the code continues after the door is closed).

**Reply by ToastyCH** [reply]
Im sure i set $392 to 8 seconds like my vfd. Im not sure what $340 is set on my machine. Will check it as soon as I can get back to my machine and the reply here.

**Reply by Dietz0r** [reply]
Just checking in from the machine. 
`$340` was indeed still set to 0. 
With `$340 = 1` it waits until the spindle reaches 20000rpm (programmed 22000rpm) and then continues.
So that is confirmed working and indeed user error on my part.
Maybe the option discription is a bit missleading and the waittime could/should be decoupled from the tolerance setting for machines without rpm feedback from the VFD (like 0-10V controlled spindles)? 
e.g. if `$340 = 0` tolerance is set wait atleast the time specified in `$394` ?

as always thanks for the quick replies @terjeio 

edit: swapped `$392` for `$394`, misstyped

**Reply by ToastyCH** [reply]
I've also checked my $340. It was also set to 0.

Now the machine waits. Even though i think the machine doesnt wait till my actual rpm is reached but maybe my "feel" is just wrong.

But like this i can work with😀

Thank you very much terjeio!

How ist this supposed here? Can I close this idea?

**Comment by S2000Stefan** [comment]
Hello ToastyCH,

> When I start a gcode like this:
>M03 S24000
>G0 X0 Y0
>G01 Z-5 F500
>My machine starts the spindle but continues imediatley. So my insert reaches the workpiece before reaching 24k rpm. (I can see the >spindle ramping up in ioSender)

To insert a delay when spinning up the spindle, please write the following line after the M3 command.

M03 S24000
**G4 P5**
G0 X0 Y0
...
Where P stands for the time in seconds. In this case 5 seconds.

Best regards
Stefan

**Reply by ToastyCH** [reply]
Hello Stefan, yeah a dealy would work. But then I always got to remember to write the command in the gcode. As I know me, Im sure i will forget it once...

If the machine would wait, that would be much "safer"for me. Or do people want to start cutting before the spindle ramped up?

**Reply by S2000Stefan** [reply]
Hello Toasty,
so i have entered this line in the postprocessor then "it" always thinks of it for me. :)
But I have heard that there is a much more elegant solution in grblHAL.
I'll take it easy then. ;)

**Reply by terjeio** [reply]
> But I have heard that there is a much more elegant solution in grblHAL.

It is available for ModBus controlled VFD spindles that can report the current RPM, or if there is an spindle encoder connected to the controller and the controller is configured to read the RPM from that. There are also a couple of drivers that can fake "spindle at speed" by ramping the PWM output (compile time option - not tested much).

**Comment by tcurdt** [comment]
So IIUC this is the logic?

`$340=1 (wait for spindle)`
`$392=12 (delay in seconds until timeout aka ALARM14)`

Would be great to have this better documented [here](https://github.com/grblHAL/core/wiki/Additional-or-extended-settings#spindle-related-settings).

**Reply by terjeio** [reply]
> Would be great to have this better documented [here](https://github.com/grblHAL/core/wiki/Additional-or-extended-settings#spindle-related-settings).

Agree. Until then you can always get at least some help from the inbuilt documentation. See the first few paragraphs [here](https://github.com/grblHAL/core/wiki/Additional-or-extended-settings#spindle-related-settings).

**Reply by tcurdt** [reply]
That's the docs I would love to see improved :)
(BTW: Happy to provide a PR to the wiki once I understand things)

It says "The timeout delay defaults to 4 seconds and is currently set from the SAFETY_DOOR_SPINDLE_DELAY symbol."

SAFETY_DOOR_SPINDLE_DELAY?

It defaults. So the timeout can be set? via `$392`?

**Reply by tcurdt** [reply]
So gsender decided to switch from grblHAL to grbl (grrr!) that's why I didn't see the inline help.
Still it would be nice to have something along these lines added to the wiki:

```
$340=10 (percent of difference in spindle speed allowed before alarm, enabled if > 0, also enables waiting for spindle on M3)
$392=12 (delay in seconds until timeout for spindle speed aka ALARM14)
```

**Comment by RetEECS** [comment]
I gather from the above discussion that (provided $340 is set > 0) other people are not seeing problems with this now?

I noticed a job where movement started before the spindle was up to speed, and found I can often reproduce the problem when I run the following g-code:
```
m3 s6000    ; spin up spindle
g0 x-200    ; perform a move that should wait until the spindle is up to speed
m5          ; stop the spindle
g0 x0       ; perform a move that should wait until the spindle has stopped
```
I have set:

```
$340=1
$392=0.0
$394=0.0
$539=0.0
```
Sometimes it waits for the spindle to spin up and down, but sometimes (particularly when spinning up) the feed starts immediately.

So far, if I perform a homing operation and then immediately run this job, the error has occurred every time. But then if run the job again it works correctly!

I added a diagnostic print to confirm that my VFD is being queried and reporting its speed correctly, which it is. The problem is just that sometimes the first g0 is held back until after the speed is right, and other times not.

Here is a screen capture showing the problem occurring:

https://github.com/user-attachments/assets/2116b58e-7ab9-4f06-a79a-d91f528e1c04

As you can see, the spindle is still spinning up as the first g0 move is taking place.

Is this just me?

**Reply by terjeio** [reply]
Which VFD driver?

**Reply by RetEECS** [reply]
Huanyang v1

**Reply by terjeio** [reply]
I've found that the $340 setting requires a controller reboot to take effect and is the likely reason. Will fix in the next commit.

**Reply by RetEECS** [reply]
My test sequence was:

1. Insert an SD card containing the new grblHAL firmware with my diagnostic printout.
2. Power on my controller to install the firmware.
3. Set all my grbl settings including $340=1.
4. Power off my controller to remove the SD card.
5. Power on the controller.
6. Perform a homing operation.
7. Load the test program.
8. Run the test

Does this count as a reboot?

Also, note that I am (almost?) always seeing the second move waiting for the spindle to stop, and that I am sometimes seeing the first move waiting for the spindle to come up to speed. If the controller were not registering the $340 setting then surely these delays would never happen?

Is there anything I can do to help diagnose this?

**Reply by terjeio** [reply]
> Does this count as a reboot?

Yes.

While testing for this issue I noticed that the "at speed" wait was randomly skipped (I use a VDF simulator so may not behave exactly as a real VFD). I only managed to "fix" this by adding a small delay before the first check for "at speed", hopefully this also applies to the real VFD. [The fix has been comitted](https://github.com/grblHAL/core/commit/2eb7179b9d1b7d96e564b8a607eac14c59302b52) along with fix for $340 requiring a reboot. The Web Builder has been updated.

**Reply by RetEECS** [reply]
Just to confirm: for my BTT OCTOPUS MAX I won't see these updates until @dresco propagates them?

**Reply by dresco** [reply]
> Just to confirm: for my BTT OCTOPUS MAX I won't see these updates until @dresco propagates them?

Hi, yes.  I'll aim to get the H7 driver caught up Monday..

**Reply by dresco** [reply]
For info, H7 driver has been updated now..

**Reply by RetEECS** [reply]
Looks good.

I have run my test 20 or 30 times and it has never skipped waiting for the spindle to reach the appropriate speed.

Thank you both for your help!
